File: app/javascript/channels/game_channel.js
import consumer from "./consumer"

document.addEventListener('turbo:load', () => {
  const element = document.getElementById('game-id');
  if (!element) return;
  const game_id = element.getAttribute('data-game-id');

  consumer.subscriptions.create({channel: "GameChannel", game_id: game_id}, {
    connected() {
      // Called when the subscription is ready for use on the server
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received(data) {
      if (data.type === 'players') {
        const el = document.getElementById('players_table');
        if (el) el.innerHTML = data.html;
      } else if (data.type === 'stories') {
        const el = document.getElementById('stories_panel');
        if (el) el.innerHTML = data.html;
      }
    }
  });
})

File: app/channels/game_channel.rb
class GameChannel < ApplicationCable::Channel
  def subscribed
    stream_from "game_channel_#{params[:game_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def self.broadcast(id, games_players)
    ActionCable.server.broadcast "game_channel_#{id}", {
      type: 'players',
      html: html(games_players, games_players.players_all_voted?)
    }
  end

  def self.broadcast_show_votes(id, games_players)
    ActionCable.server.broadcast "game_channel_#{id}", {
      type: 'players',
      html: html(games_players, true)
    }
  end

  def self.broadcast_stories(game)
    ActionCable.server.broadcast "game_channel_#{game.id}", {
      type: 'stories',
      html: stories_html(game)
    }
  end

  def self.html(players, show_player_votes)
    ApplicationController.render(
      partial: 'games/players_table',
      locals: { players: players, :show_player_votes => show_player_votes }
    )
  end

  def self.stories_html(game)
    ApplicationController.render(
      partial: 'games/stories_panel',
      locals: { game: game, stories: game.stories.ordered }
    )
  end
end
